File: 2024/20/solution.rb
# frozen_string_literal: true

require_relative '../../lib/aoc'
input = $stdin.read.strip

grid = AOC::Cgrid.new(input)
# puts grid.display

start = grid.find_one 'S'
target = grid.find_one 'E'

def solve(grid, start, target)
  dirs = grid.dir4
  scores = { start => 0 }
  heads = Queue.new
  heads.push start
  while scores[target].nil? && heads.size.positive?
    head = heads.pop
    score = scores[head]
    dirs.each do |dir|
      nexthead = head + dir
      next if grid[nexthead].nil?
      next if grid[nexthead] == '#'
      next unless scores[nexthead].nil?

      heads.push nexthead
      scores[nexthead] = score + 1
    end
  end
  scores
end

def cheats(grid, start, target, cheattime)
  imps = {}
  imps.default = 0
  scores = solve(grid, start, target)
  cleanscore = scores[target]
  racetrack = grid.find_all 'SE.'.chars
  racetrack = racetrack.sort_by { |p| scores[p] }
  racetrack.freeze

  # verify that the racetrack has no dead ends or wider parts
  # also ensure there are no longer paths from S->E
  # This means that every point has a proper score when solving with BFS once
  raise 'Racetrack assumption not correct' if cleanscore != racetrack.size - 1

  (0..racetrack.size - 1).each do |idx|
    p1 = racetrack[idx]

    # Only consider within the maximum cheattime (manhattan) range.
    # Other points are not valid cheats
    diststart = scores[p1]
    (-cheattime..cheattime).each do |r|
      left = cheattime - r.abs
      (-left..left).each do |im|
        p2 = p1 + r + im.i
        next if grid[p2] == '#'

        dist = grid.manh(p1, p2)
        next if dist > cheattime

        distend = cleanscore - scores[p2]
        newdist = diststart + distend + dist
        delta = cleanscore - newdist
        imps[delta] += 1 if delta.positive?
      end
    end
  end
  imps
end

# p cheats(grid,start,target,2)
# p cheats(grid,start,target,20).filter {|k,v| k >= 50}

p [cheats(grid, start, target, 2).filter { |k, _| k >= 100 }.values.sum,
   cheats(grid, start, target, 20).filter { |k, _| k >= 100 }.values.sum]

File: 2024/20/solution.py
from collections import deque,Counter

grid = [l.strip() for l in open(0)]
rows = len(grid)
cols = len(grid[0])
start = None
end = None
racetrack = []

for r in range(rows):
    for c in range(cols):
        p = (r,c)
        v = grid[r][c]
        if v == '#': continue
        racetrack.append(p)
        if v == 'E': 
            end = p
        if v == 'S':
            start = p

# def manh(p1,p2): r1,c1 = p1 ; r2,c2 = p2 ; return abs(r1-r2) + abs(c1-c2)

def scores(start,end):
    dirs = [[0,1],[0,-1],[1,0],[-1,0]]
    scores = {}
    scores[start] = 0
    heads =  deque([start])

    while not end in scores and heads:
        r,c = heads.popleft()
        # print(r,c)
        for dr,dc in dirs:
            nr, nc = r+dr,c+dc
            if nr < 0 or nr >= rows: continue
            if nc < 0 or nc >= cols: continue
            if (nr,nc) in scores: continue
            if grid[nr][nc] == '#': continue
            scores[(nr,nc)] = scores[(r,c)]+1
            heads.append((nr,nc))
       
    return scores

def cheats(cheattime,threshold):
    imps = Counter()
    tracksize = len(racetrack)
    for idx in range(tracksize-1):
        p1 = racetrack[idx]
        x,y = p1
        startscore = sc[p1]
        for dx in range(-cheattime,cheattime+1):
            rest = cheattime - abs(dx)
            for dy in range(-rest,rest+1):
                chp = (x+dx,y+dy)
                if chp in sc:
                    delta_sc = sc[chp]-startscore-abs(dx)-abs(dy)
                    if delta_sc >= threshold: imps[delta_sc]+= 1
    
    return imps

sc = scores(start,end)

# racetrack.sort(key=lambda t: sc[t])

part1 = sum(cheats(2,100).values())
part2 = sum(cheats(20,100).values())

print(part1, part2)


